crates/ruff_linter/src/rules/flake8_pyi/rules/bad_generator_return_type.rs

```diff
@@ -176,16 +214,151 @@ pub(crate) fn bad_generator_return_type(
             }
         }
     };
-
-    checker.diagnostics.push(Diagnostic::new(
+    let mut diagnostic = Diagnostic::new(
         GeneratorReturnFromIterMethod {
-            better_return_type: better_return_type.to_string(),
-            method_name: name.to_string(),
+            return_type: member.to_iter(),
+            method,
         },
         function_def.identifier(),
-    ));
+    );
+    if let Some(fix) = generate_fix(
+        function_def,
+        returns,
+        yield_type_info,
+        module,
+        member,
+        checker,
+    ) {
+        diagnostic.set_fix(fix);
+    };
+    checker.diagnostics.push(diagnostic);
 }
 
+/// Returns `true` if the [`ast::Expr`] is a `None` literal or a `typing.Any` expression.
 fn is_any_or_none(expr: &ast::Expr, semantic: &SemanticModel) -> bool {
-    semantic.match_typing_expr(expr, "Any") || matches!(expr, ast::Expr::NoneLiteral(_))
+    expr.is_none_literal_expr() || semantic.match_typing_expr(expr, "Any")
+}
+
+/// Generate a [`Fix`] to convert the return type annotation to `Iterator` or `AsyncIterator`.
+fn generate_fix(
+    function_def: &ast::StmtFunctionDef,
+    returns: &ast::Expr,
+    yield_type_info: Option<YieldTypeInfo>,
+    module: Module,
+    member: Generator,
+    checker: &Checker,
+) -> Option<Fix> {
+    let expr = map_subscript(returns);
+
+    let (import_edit, binding) = checker
+        .importer()
+        .get_or_import_symbol(
+            &ImportRequest::import_from(&module.to_string(), &member.to_iter().to_string()),
+            expr.start(),
+            checker.semantic(),
+        )
+        .ok()?;
+    let binding_edit = Edit::range_replacement(binding, expr.range());
+    let yield_edit = yield_type_info.map(|yield_type_info| {
+        Edit::range_replacement(
+            checker.generator().expr(yield_type_info.expr),
+            yield_type_info.range(),
+        )
+    });
+
+    // Mark as unsafe if it's a runtime Python file and the body has more than one statement in it.
+    let applicability = if checker.source_type.is_stub() || function_def.body.len() == 1 {
+        Applicability::Safe
+    } else {
+        Applicability::Unsafe
+    };
+
+    Some(Fix::applicable_edits(
+        import_edit,
+        std::iter::once(binding_edit).chain(yield_edit),
+        applicability,
+    ))
+}
+
+#[derive(Debug)]
+struct YieldTypeInfo<'a> {
+    expr: &'a ast::Expr,
+    range: TextRange,
+}
+
+impl Ranged for YieldTypeInfo<'_> {
+    fn range(&self) -> TextRange {
+        self.range
+    }
+}
```

Comment:
For my education: what's the advantage of implementing the trait here rather than just reading the attribute?

Reply:
It's really minor, more of a consistency thing... If you implement `Ranged`, then you get methods like `.start()` and `.end()` instead of having to do `.range.start()`.